keps/sig-apps/0753-sidecarcontainers.md

```diff
@@ -315,8 +1348,214 @@ Older Kubelets should still be able to schedule Pods that have sidecar container
+ordering:
+ * Blackhole all the traffic until service mesh container is up (usually using
+   an initContainer for this)
+ * Some trickery (sleep preStop hooks or some alternative) to not be killed
+   before other containers that need to use the network. Otherwise, traffic for
+   those containers will be blackholed
+
+This means that if the app container is started before the service mesh is
+started and ready, all traffic will be blackholed and the app needs to retry.
+Once the service mesh container is ready, traffic will be allowed.
+
+This has another major disadvantage: several apps crash if traffic is blackholed
+during startup (common in some rails middleware, for example) and have to resort
+to some kind of workaround, like [this one][linkerd-wait] to wait. This makes
+also service mesh miss their goal of augmenting containers functionality without
+modifying the main application.
+
+Istio has an alternative to the initContainer hack. Istio [has an
+option][istio-cni-opt] to integrate with CNI and inject the blackhole from there
+instead of using the initContainer. In that case, it will do (just c&p from the
+link, in case it breaks in the future):
+
+> By default Istio injects an initContainer, istio-init, in pods deployed in the mesh. The istio-init container sets up the pod network traffic redirection to/from the Istio sidecar proxy. This requires the user or service-account deploying pods to the mesh to have sufficient Kubernetes RBAC permissions to deploy containers with the NET_ADMIN and NET_RAW capabilities. Requiring Istio users to have elevated Kubernetes RBAC permissions is problematic for some organizations’ security compliance
+> ...
+> The Istio CNI plugin performs the Istio mesh pod traffic redirection in the Kubernetes pod lifecycle’s network setup phase, thereby removing the requirement for the NET_ADMIN and NET_RAW capabilities for users deploying pods into the Istio mesh. The Istio CNI plugin replaces the functionality provided by the istio-init container.
+
+In other words, Istio has an alternative to configure the traffic blockhole
+without an initContainer. But the other problems and hacks mentioned remain,
+though.
+
+[linkerd-last-container]: https://github.com/linkerd/linkerd2/issues/4758#issuecomment-658457737
+[istio-cni-opt]: https://istio.io/latest/docs/setup/additional-setup/cni/
+[linkerd-wait]: https://github.com/olix0r/linkerd-await
+
+##### Istio bug report
+
+There is also a [2 years old bug][istio-bug-report] from Istio devs that this
+KEP will fix. In addition, similar benefit is expected for Linkerd, as we talked
+with Linkerd devs.
+
+One of the things mentioned there is that, at least in 2018, a workaround used
+was to tell the user to run a script to wait for the service mesh to start on
+their containers.
+
+Rodrigo will reach out to Istio devs to see if the situation changed since 2018.
+
+[istio-bug-report]: https://github.com/kubernetes/kubernetes/issues/65502
```

Comment:
@howardjohn one last question to address this ^ section, it it is not too much to ask.

Do you still need to use, as the bug says, some scripts for other applications in the pod to wait for the service mesh to be ready? Or was the situation changed a lot in these 2 years? 

If it changed a lot and you can share your pain points and which ones this KEP effectively address, it would be *great*.

Reply:
cc @nmittler that opened the bug, maybe? :)